bcda/logging/middleware.go

```diff
@@ -0,0 +1,86 @@
+package logging
+
+import (
+	"fmt"
+	"net/http"
+	"time"
+
+	jwt "github.com/dgrijalva/jwt-go"
+	"github.com/go-chi/chi/middleware"
+	"github.com/sirupsen/logrus"
+)
+
+// https://github.com/go-chi/chi/blob/master/_examples/logging/main.go
+
+func NewStructuredLogger() func(next http.Handler) http.Handler {
+	logger := logrus.New()
+	logger.Formatter = &logrus.JSONFormatter{}
+	return middleware.RequestLogger(&StructuredLogger{Logger: logger})
+}
+
+type StructuredLogger struct {
+	Logger *logrus.Logger
+}
+
+func (l *StructuredLogger) NewLogEntry(r *http.Request) middleware.LogEntry {
+	entry := &StructuredLoggerEntry{Logger: logrus.NewEntry(l.Logger)}
+	logFields := logrus.Fields{}
+
+	logFields["ts"] = time.Now().UTC().Format(time.RFC1123)
+
+	if reqID := middleware.GetReqID(r.Context()); reqID != "" {
+		logFields["req_id"] = reqID
+	}
+
+	scheme := "http"
+	if r.TLS != nil {
+		scheme = "https"
+	}
+	logFields["http_scheme"] = scheme
+	logFields["http_proto"] = r.Proto
+	logFields["http_method"] = r.Method
+
+	logFields["remote_addr"] = r.RemoteAddr
+	logFields["user_agent"] = r.UserAgent()
+
+	logFields["uri"] = fmt.Sprintf("%s://%s%s", scheme, r.Host, r.RequestURI)
+
+	token := r.Context().Value("token")
+	if token != nil {
+		token := token.(*jwt.Token)
+		claims := token.Claims.(jwt.MapClaims)
+		logFields["aco"] = claims["aco"]
+		logFields["sub"] = claims["sub"]
+	}
+
+	hashedToken := r.Context().Value("hashedToken")
+	if hashedToken != nil {
+		logFields["token"] = hashedToken.(string)
```

Comment:
What do you think about just using the token UUID here now that we have that? Or maybe there is an advantage to having both?